lib/sms/sms.c

```diff
@@ -0,0 +1,322 @@
+{
+	if (cmt_rsp.alpha != NULL) {
+		k_free(cmt_rsp.alpha);
+	}
+
+	if (cmt_rsp.pdu != NULL) {
+		k_free(cmt_rsp.pdu);
+	}
+
+	/* Save alpha as a null-terminated String. */
+	size_t alpha_len;
+	(void)at_params_size_get(&resp_list, 1, &alpha_len);
+
+	cmt_rsp.alpha = k_malloc(alpha_len + 1);
+	if (cmt_rsp.alpha == NULL) {
+		return -ENOMEM;
+	}
+	(void)at_params_string_get(&resp_list, 1, cmt_rsp.alpha, &alpha_len);
+	cmt_rsp.alpha[alpha_len] = '\0';
+
+	/* Length field saved as number. */
+	(void)at_params_short_get(&resp_list, 2, &cmt_rsp.length);
+
+	/* Save PDU as a null-terminated String. */
+	size_t pdu_len;
+	(void)at_params_size_get(&resp_list, 3, &pdu_len);
+	cmt_rsp.pdu = k_malloc(pdu_len + 1);
+	if (cmt_rsp.pdu == NULL) {
+		return -ENOMEM;
+	}
+
+	(void)at_params_string_get(&resp_list, 3, cmt_rsp.pdu, &pdu_len);
+	cmt_rsp.pdu[pdu_len] = '\0';
+
+	return 0;
+}
+
+/** @brief Handler for AT responses and unsolicited events. */
+void sms_at_handler(void *context, char *at_notif)
+{
+	ARG_UNUSED(context);
+	LOG_DBG("AT notification '%s'", log_strdup(at_notif));
+
+	/* Check if the notification is a CMT event. */
+	if (!sms_is_cmd_notification(at_notif)) {
+		LOG_DBG("Ignore notifications other than CMT");
+		return;
+	}
+
+	/* Parse and validate the CMT notification, then extract parameters. */
+	if (sms_cmt_notif_parse(at_notif) != 0) {
+		LOG_ERR("Invalid CMT notification");
+		return;
+	}
+
+	/* Extract and save the SMS notification parameters. */
+	int valid_notif = sms_cmt_notif_save();
+
+	/* Ack the SMS PDU in any case. */
+	int ret = at_cmd_write(AT_SMS_PDU_ACK, NULL, 0, NULL);
+
+	if (ret != 0) {
+		LOG_ERR("Unable to ACK the SMS PDU");
+	}
+
+	if (valid_notif != 0) {
+		LOG_ERR("Invalid SMS notification format");
+		return;
+	}
+
+	/* Notify all subsribers. */
+	LOG_DBG("Valid SMS notification decoded");
+	for (sms_handle_t i = 0; i < ARRAY_SIZE(subsribers); i++) {
+		const struct sms_subscriber *s = subsribers + i;
+
+		if (s->listener != NULL) {
+			s->listener(&cmt_rsp, s->ctx);
+		}
+	}
+}
+
+int sms_init(void)
+{
+	int ret = at_params_list_init(&resp_list, AT_SMS_PARAMS_COUNT_MAX);
+
+	if (ret) {
+		LOG_ERR("AT params error, err: %d", ret);
+		return ret;
+	}
+
+	/* Check if one SMS client has already been registered. */
+	ret = at_cmd_write(AT_SMS_SUBSCRIBER_READ, resp, sizeof(resp), NULL);
+	if (ret) {
+		LOG_ERR("Unable to check if an SMS client exists, err: %d",
+			ret);
+		return ret;
+	}
+
+	ret = at_parser_max_params_from_str(resp, NULL, &resp_list,
+					    AT_CNMI_PARAMS_COUNT);
+	if (ret) {
+		LOG_INF("%s", log_strdup(resp));
+		LOG_ERR("Invalid AT response, err: %d", ret);
+		return ret;
+	}
+
+	int cnmi = 0, value;
+
+	for (int i = 1; i < AT_CNMI_PARAMS_COUNT - 1; i++) {
+		ret = at_params_int_get(&resp_list, i, &value);
+		if (ret) {
+			LOG_ERR("Invalid AT response parameters, err: %d", ret);
+			return ret;
+		}
+		cnmi += value;
+	}
+
+	/* No client are registered when the first 4 parameters are 0. */
```

Comment:
The comment says "the first 4", but we are checking 6.

Reply:
I am reading the parameter indexes [1..4].